maven/impl-maven-embedded/src/main/java/org/jboss/shrinkwrap/resolver/impl/maven/embedded/DistributionStageImpl.java

```diff
@@ -80,19 +83,31 @@ public boolean accept(File file) {
         return extracted[0];
     }
 
+    private File createTempFile(File toExtract) {
+        try {
+            final File tempFile = Files.createTempFile(toExtract.toPath(), "temp", UUID.randomUUID().toString()).toFile();
+            tempFile.deleteOnExit();
+
+            return tempFile;
+        } catch (IOException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
     private File extract(File downloaded, String downloadedZipMd5hash) {
         File withExtractedDir = checkIfItIsAlreadyExtracted(downloadedZipMd5hash);
```

Comment:
if the zip is not already extracted, but the extraction is being processed (the marker file is present), then you don't wait until the file is deleted but start another extraction which could cause some problems - two processes are extracting to the same directory. You should wait until the file is removed and then return the path to the directory. Some timeout would make sense as well.